### PR TITLE
Enable dtype flexibility

### PR DESCRIPTION
**Pull request by TimFelixBeyer**
First off, I apologize for the noisy diff, I am not very experienced with GitHub PRs. 
Let me know if there are any changes you'd like me to make.

This PR enables the user to choose between tf.float32 and tf.float64.
This can be done by setting `tf.keras.backend.set_float(dtype)`, and by then providing the inputs with the same dtype.

The accuracy of the computation will vary with the choice of dtype. 
I conducted some tests (including gradient computation) and these are my results: 
adjoint_float_test_rtol_error
Note: Because the error estimation of the adaptive solver gets worse, f32 doesn't always improve speed and (especially when choosing tolerances below 1e-7) will actually take longer than f64 in some cases.

As they should not be needed anymore I have removed `cast_double` and `func_cast_double`.
There are quite a few more minor changes that I have included, like spelling errors/trailing whitespace/code cleanup etc. 

All of the tests pass on my machine.

### CODE REVIEW

**Comment by titu1994** [comment]
This looks great, thank you very much for this PR! I see there are some merge conflicts, if you could rebase with the master branch and correct these files, i could go ahead and merge this.

**Comment by TimFelixBeyer** [comment]
I am unsure about the reason for the current failure, the way I addressed it on my machine was by changing some tensorflow-internal lines. I have looked around the code quite a bit but I don't know which change introduces the fail.

**Comment by titu1994** [comment]
Yes I noticed in your fork that to mention edits to core tensorflow. That's not something we can reasonably expect a user to be willing to do.

**Comment by titu1994** [comment]
Mainly all of the adjoint tests break due to this. Let me look into the code a little more over the day and I'll get back to you if I can find a solution.

**Comment by TimFelixBeyer** [comment]
I have found a fix :)

**Comment by titu1994** [comment]
Fantastic ! Thank for this and sorry for the delay. This week has been hectic

File: tfdiffeq/adjoint.py [hunk 1]
```diff
@@ -182,18 +175,18 @@ def augmented_dynamics(t, y_aug):
         adj_params_splits = tf.split(adj_params, var_flat_lens)
         adj_params_list = [tf.reshape(p, v_shape)
                            for p, v_shape in zip(adj_params_splits, var_shapes)]
-
         return (*adj_y, time_vjps), adj_params_list
 
     return ans, grad
 
 
-def odeint_adjoint(func, y0, t, rtol=1e-6, atol=1e-12, method=None, options=None, adjoint_method=None, adjoint_rtol=None,
```
Comment (titu1994) [inline tfdiffeq/adjoint.py]: These default tolerance should be reverted to be comparable with torch diff eq

File: tfdiffeq/adjoint.py [hunk 2]
```diff
@@ -207,33 +200,29 @@ def odeint_adjoint(func, y0, t, rtol=1e-6, atol=1e-12, method=None, options=None
...
-
-        # build the function to get its variables
-        # if not func.built:
-        #     _ = func(t, y0)
-
-        global _arguments
-        _arguments = _Arguments(func, method, options, rtol, atol,
-                                adjoint_method, adjoint_rtol, adjoint_atol, adjoint_options)
+    # build the function to get its variables
+    if not func.built:
```
Comment (titu1994) [inline tfdiffeq/adjoint.py]: There's no need for this, and it crashes certain cases where this assumption of zero t fails

**Review by titu1994 — APPROVED** [review]
Minor changes, but overall amazing work

**Comment by TimFelixBeyer** [comment]
The tests were broken due to the newly released TF 2.3.0 . 
I have included a possible fix.

**Comment by titu1994** [comment]
Perfect. Merged. Thank you once again for this awesome PR ! It's really very useful to not have explicit casts everywhere.
